### PR TITLE
Uncertainty fixes and user experience improvement

### PR DESCRIPTION
**Pull request by mliu49**
### Motivation or Problem
This PR contains 3 parts:
- Minor maintenance of uncertainty analysis functionality to make sure it works
- Major user experience improvements in how uncertainty analysis is performed
- Updates to related IPython notebooks and examples

### Description of Changes
User experience improvements:
- It is now possible to request uncertainty analysis be automatically performed at the end of an RMG job via input file settings
- The `simulate.py` script can be similarly used to run uncertainty analysis automatically
- Uncertainty values for uncorrelated global analysis can be retrieved automatically instead of manually entered

IPython notebook updates:
- findParameterSourcesAndAssignUncertainties.ipynb
- localUncertainty.ipynb
- globalUncertainty.ipynb
- canteraSimulation.ipynb
- canteraSensitivityComparison.ipynb

There are also a number of bug fixes. Documentation updates to be included soon.

### Testing
The input files in `ipython/data/ethane_model` and `ipython/data/pdd_model` can be run to test uncertainty analysis following an RMG job. All of the IPython notebooks should work. Note that globalUncertainty does require setting up MUQ properly, which is non-trivial currently.

To set up MUQ:
- I suggest creating a new environment first
- Add `/home/user/anaconda2/envs/rmg_new/lib` to path
- Do `conda install hdf5=1.8`
- Do `conda install icu=54`

### Reviewer Tips
I suggest going through the commits in order rather than viewing all changes at once.

### CODE REVIEW

File: rmgpy/tools/uncertainty.py [hunk 1]
```diff
@@ -708,7 +708,17 @@ def localAnalysis(self, sensitiveSpecies, correlated=False, number=10, fileforma
...
-            rPath = os.path.join(self.outputDirectory, 'kineticsLocalUncertainty_{0}'.format(sensSpecies.toChemkin()) + fileformat)
+            if correlated:
+                folder = os.path.join(self.outputDirectory, 'correlated')
+            else:
+                folder = os.path.join(self.outputDirectory, 'uncorrelated')
+            if not os.path.exists(folder):
+                try:
+                    os.makedirs(folder)
+                except:
+                    raise Exception('Uncertainty output directory could not be created.')
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/uncertainty.py]: Fixed. I forget why I decided to catch this exception at all. I might have just copied it from somewhere.

File: rmgpy/tools/muq.py [hunk 2]
```diff
@@ -81,6 +81,29 @@ def __init__(self, cantera, outputSpeciesList, kParams, kUncertainty, gParams, g
         # The size of the uncertain inputs: [parameters affecting k, parameters affecting free energy G]         
         self.inputSize = [len(kParams) + len(gParams)]
 
+        if not self.correlated:
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/muq.py]: I'm not sure. Ideally, we should update the MUQ binaries so none of this is an issue anymore. Idk when I'll get a chance to look into doing so though.

File: rmgpy/tools/muq.py [hunk 3]
```diff
@@ -476,8 +476,8 @@ def analyzeResults(self):
                             description = 'dln[{0}]/dln[{1}]'.format(outputSpecies.toChemkin(), descriptor)
 
                         print '{0:55} {1:10.3f} {2:10.3f}'.format(description,
-                                                                  mainSens[outputIndex][parameterIndex],
-                                                                  totalSens[outputIndex][parameterIndex])
+                                                                  100*mainSens[outputIndex][parameterIndex],
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/muq.py]: Fixed

File: rmgpy/tools/muq.py [hunk 4]
```diff
@@ -476,8 +476,8 @@ def analyzeResults(self):
                             description = 'dln[{0}]/dln[{1}]'.format(outputSpecies.toChemkin(), descriptor)
 
                         print '{0:55} {1:10.3f} {2:10.3f}'.format(description,
-                                                                  mainSens[outputIndex][parameterIndex],
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/muq.py]: I had put `(%)` in the column headers, but I've moved them to the values to make it clearer.

File: rmgpy/rmg/input.py [hunk 5]
```diff
@@ -675,6 +675,20 @@ def thermoCentralDatabase(host,
                                                             application)
                     
 
+def uncertainty(localAnalysis=False, globalAnalysis=False, uncorrelated=True, correlated=True,
+                localNumber=10, globalNumber=5, terminationTime=None, pceRunTime=1800):
+    rmg.uncertainty = {
+        'local': localAnalysis if not globalAnalysis else True,  # Must run local before global
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/rmg/input.py]: Added a logging statement.

File: rmgpy/tools/uncertainty.py [hunk 6]
```diff
@@ -465,40 +469,43 @@ def compileAllSources(self):
...
             self.allKineticSources['Training'][familyLabel] = list(allKineticSources['Training'][familyLabel])
-    
-    def assignParameterUncertainties(self, gParamEngine = ThermoParameterUncertainty(), kParamEngine = KineticParameterUncertainty(), correlated=False):
+
+    def assignParameterUncertainties(self, gParamEngine=None, kParamEngine=None, correlated=False):
         """
         Assign uncertainties based on the sources of the species thermo and reaction kinetics.
         """
-        
+        if gParamEngine is None:
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/uncertainty.py]: Haha, this was manually changed.

File: rmgpy/tools/plot.py [hunk 7]
```diff
@@ -96,12 +96,12 @@ def parseCSVData(csvFile):
             rxn = data.label.split()[1]
             index = data.label.split()[0][:-2].rpartition('dln[k')[2]
             data.reaction = rxn
-            data.index = index
+            data.index = int(index)
```
[the remark this answers is not in the document]
Comment (mliu49) [reply rmgpy/tools/plot.py]: It was a string, parsed from the csv.

**Comment by mliu49** [comment]
@amarkpayne Would you still like to review this?

@alongd Could you take a look at the fixups and documentation additions?

File: documentation/source/users/rmg/input.rst [hunk 8]
```diff
@@ -771,6 +771,70 @@ to turn off pressure dependence for all molecules larger than the given number
...
+The ``uncorrelated`` and ``correlated`` options refer to two approaches for uncertainty analysis. Uncorrelated means
+that all input parameters are considered to be independent, each with their own uncertainty bounds. Thus, the output
+uncertainty distribution is determined on the basis that every input parameter could vary within the full range of
+its uncertainty bounds. Correlated means that inherent relationships between parameters (such as rate rules for kinetics
+or group additivity values for thermochemistry) are accounted for, which reduces the uncertainty space of the input
+parameters.
+
+Finally, there are a few miscellaneous options for global uncertainty analysis. The ``terminationTime`` applies for the
+reactor simulation. It is only necessary if termination time is not specified in the reactor settings (i.e. only other
+termination criteria are used). The ``pceRunTime`` sets the time limit for fitting the PCE to the output surface.
```
Comment (alongd) [inline documentation/source/users/rmg/input.rst]: Define PCE?

Comment (mliu49) [reply documentation/source/users/rmg/input.rst]: I added the abbreviation to the first mention of Polynomial Chaos Expansion earlier in the section (line 806).

**Comment by mliu49** [comment]
I rebased again. Please go ahead and merge whenever you're ready.

**Comment by alongd** [comment]
Thanks for this awesome PR!

**Pull request merged**